opal/mca/btl/base/btl_base_am_rdma.c

```diff
@@ -0,0 +1,1064 @@
+    descriptor->des_context = context;
+
+    hdr = (mca_btl_base_rdma_hdr_t *) descriptor->des_segments[0].seg_addr.pval;
+    hdr->type = type;
+
+    if (!mca_btl_base_rdma_is_atomic (type)) {
+        hdr->data.rdma.use_rdma = use_rdma;
+    } else {
+        hdr->data.atomic.op = op;
+        hdr->data.atomic.operand[0] = operand1;
+        hdr->data.atomic.operand[1] = operand2;
+    }
+
+    hdr->context = (uintptr_t) context;
+
+    if (use_rdma && btl->btl_register_mem) {
+        uint8_t *handle_buffer = (uint8_t *) (hdr + 1);
+        memcpy (handle_buffer, local_handle, btl->btl_registration_handle_size);
+        handle_buffer += btl->btl_registration_handle_size;
+        memcpy (handle_buffer, remote_handle, btl->btl_registration_handle_size);
+    }
+
+    return mca_btl_base_am_rdma_advance (btl, endpoint, context, /*send_descriptor=*/true);
+}
+
+static mca_btl_base_rdma_operation_t *mca_btl_base_rdma_alloc_operation (mca_btl_base_module_t *btl,
+                                                                         struct mca_btl_base_endpoint_t *endpoint,
+                                                                         mca_btl_base_descriptor_t *descriptor,
+                                                                         const mca_btl_base_rdma_hdr_t *hdr)
+{
+    mca_btl_base_rdma_operation_t *operation = OBJ_NEW(mca_btl_base_rdma_operation_t);
+    if (NULL == operation) {
+        return NULL;
+    }
+
+    operation->btl = btl;
+    operation->endpoint = endpoint;
+    operation->descriptor = descriptor;
+    operation->is_completed = false;
+    operation->is_queued = false;
+    memcpy (&operation->hdr, hdr, sizeof (*hdr));
+
+    if (!mca_btl_base_rdma_is_atomic (hdr->type) && hdr->data.rdma.use_rdma && btl->btl_register_mem) {
+        const uint8_t *handle_data = (const uint8_t *) (hdr + 1);
+        /* the initiator packs these in order of their local and then remote. */
+        memcpy (operation->remote_handle_data, handle_data, btl->btl_registration_handle_size);
+        handle_data += btl->btl_registration_handle_size;
+        memcpy (operation->local_handle_data, handle_data, btl->btl_registration_handle_size);
+    }
+
+    return operation;
+}
+
+static void mca_btl_base_rdma_queue_operation (mca_btl_base_module_t *btl, struct mca_btl_base_endpoint_t *endpoint,
+                                               mca_btl_base_descriptor_t *descriptor, uint64_t atomic_response,
+                                               const mca_btl_base_rdma_hdr_t *hdr, mca_btl_base_rdma_operation_t *operation) {
+    if (NULL == operation) {
+        operation = mca_btl_base_rdma_alloc_operation (btl, endpoint, descriptor, hdr);
+        if (NULL == operation) {
+            /* couldn't even allocate a small amount of memory. not much else can be done. */
+            BTL_ERROR(("could not allocate memory to queue active-message RDMA operation"));
+            abort();
```

Comment:
Is there a more graceful way then to call abort here?

Reply:
Yes, please do not call `abort()` -- that can lead to terrabytes of corefiles.  ☹️

Reply:
It is either abort or SEGV. IMHO abort is much more graceful. If we get here things are bad and we will abort or SEGV somewhere if we can't even allocate 64 bytes.

Reply:
`abort()` drops a corefile -- that's kinda anti-social.

If you need to exit, please call the btl error function.

Reply:
Already called. In that case abort is not even reached.

Reply:
Wait. BTL_ERROR doesn't exit so we still want it to exit here. It will SEGV which will dump core files.

Reply:
So, I can remove the aborts() but the semantics of this error will not change. It will print the error message then crash when operation is dereferenced.

Reply:
I'm not sure I agree with @jsquyres on the abort thing; I think this is precisely when we want a core file (and, I believe, we call abort elsewhere).  This is a should "never, ever happen" type thing, not a "we got our configuration wrong" type thing.

Reply:
Maybe we should have an option for dropping corefiles.  It is very anti-social to have an MTT run fill up filesystems with terrabytes of corefiles.

Reply:
Shouldn't happen ever in MTT until you are expecting to completely run out of memory. But, again, in that case we really can't control whether you core or not. Up to your shell to set ```ulimit -c 0```.